peachjam/views/generic_views.py

```diff
@@ -218,16 +218,22 @@ def fetch_docs(self, works):
         return result
 
     def add_relationships(self, context):
-        context["relationships_as_subject"] = rels_as_subject = list(
-            Relationship.for_subject_document(context["document"])
-            .prefetch_related("subject_work", "object_work")
-            .select_related("predicate")
-        )
-        context["relationships_as_object"] = rels_as_object = list(
-            Relationship.for_object_document(context["document"])
-            .prefetch_related("object_work", "subject_work")
-            .select_related("predicate")
-        )
+        relationships_as_subject = [
+            relationship
+            for relationship in Relationship.for_subject_document(context["document"])
+            if relationship.object_work
+            and CoreDocument.objects.filter(work=relationship.object_work)
```

Comment:
This will call the database once for every relationship, which is very inefficient. You generally want to do cross-object filtering like this in the database.

Try something like this instead:

```python
Relationship.for_subject_document(context["document"])
  .filter(subject_work__documents__isnull=False)
  .distinct('pk')
```

when we join `subject_work` to `documents` we could have multiple results, so we need to add the `distinct` to ensure we get unique relationships only.